# Assignment 01


<p style='text-align: right; '> Kristine Martinez </p>
<p style='text-align: right; '> 8/27/2022 </p>

**For this assignment I have chosen to write about Newton's Method. I will show the algorithm of the method, give an example using Python, and discuss the advantages and disadvantages of the method.** 
_____________________________

Newton's method is an iterative process for finding successive roots of a real-valued function, $f(x)=0$. Given an function $f$ defined over the real numbers $x$ and the function's derivative $f'$, we start with an initial guess $x_0$ for a root of the funtion. The function must satisfy all assumptions made in the derivation of the formula and the initial guess is close to the solution. The algorithm of Newton's method begins as is:
1. First we choose an initial guess $x_0$
2. Then starting with initial guess $x_0$ we compute a better approximation $x_1$ $$x_1=x_0-\frac{f(x_0)}{f'(x_0)}\quad where\ f'(x_0)\neq0$$
3. The process is repeated until a successive value is achieved for the desired number of decimal places given by the formula $$x_{n+1}=x_n-\frac{f(x_n)}{f'(x_n)}$$ 

We use this algorithm successively until $x_n$ coverges to the solution $x^*$.

Now let's use Newton's method for the following example: $$f(x)=x^2-5$$ First, let us define the function $f$ and the function's derivative $f'$:

In [87]:
import sympy as sy
import numpy as np

x=sy.Symbol('x')

f=x**2-5
diff_f=sy.diff(f,x)

f_x=sy.lambdify(x,f)
diff_f_x=sy.lambdify(x,diff_f)

In our example, the function $f$ has the solution $x^*=2.2360679775$. 

Now we will implement Newton's method by creating a function and go through 10 iterations of the algorithm.

In [88]:
def newtonsmethod(x0,iterations,f,df):
    x=x0
    for i in range(iterations):
        x=x-f(x)/df(x)
    return x

In [89]:
approximation = newtonsmethod(2,10,f_x,diff_f_x)
approximation

2.23606797749979

With Newton's method, the root of the function $f(x)=x^2-5$ is approximately 2.23606797749979 since convergence occurs to that value over the number of iterations.

## Conclusion

The advantages of Newton's method is its fast. Although the method works fast, some disadvantages are that our initial guess $x_0$ must be close to the solution and we need to compute the derivative of the function $f'(x)$, which can take a lot of computation and time.